# Paul Kelly
# DATS 6202
# Homework 1
# 6/20/23

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Overview" data-toc-modified-id="Overview-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Overview</a></span></li><li><span><a href="#Notebook-Configuration" data-toc-modified-id="Notebook-Configuration-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Notebook Configuration</a></span><ul class="toc-item"><li><span><a href="#Google-drive" data-toc-modified-id="Google-drive-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Google drive</a></span></li><li><span><a href="#Warning" data-toc-modified-id="Warning-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Warning</a></span></li><li><span><a href="#Matplotlib" data-toc-modified-id="Matplotlib-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Matplotlib</a></span></li><li><span><a href="#TensorFlow" data-toc-modified-id="TensorFlow-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>TensorFlow</a></span></li><li><span><a href="#Random-seed" data-toc-modified-id="Random-seed-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Random seed</a></span></li></ul></li><li><span><a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Preprocessing</a></span></li><li><span><a href="#Hyperparameter-Tuning" data-toc-modified-id="Hyperparameter-Tuning-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Hyperparameter Tuning</a></span></li><li><span><a href="#Model-Selection" data-toc-modified-id="Model-Selection-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Model Selection</a></span></li><li><span><a href="#Generating-the-Submission-File" data-toc-modified-id="Generating-the-Submission-File-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Generating the Submission File</a></span><ul class="toc-item"><li><span><a href="#Creating-the-directory-for-the-submission-file" data-toc-modified-id="Creating-the-directory-for-the-submission-file-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Creating the directory for the submission file</a></span></li><li><span><a href="#Generating-the-submission-file" data-toc-modified-id="Generating-the-submission-file-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Generating the submission file</a></span></li></ul></li></ul></div>

<b>
<p>
<center>
<font size="5">
Popular Machine Learning Methods: Idea, Practice and Math
</font>
</center>
</p>

<p>
<center>
<font size="4">
Part 2, Chapter 2, Section 2: Training Shallow Models
</font>
</center>
</p>

<p>
<center>
<font size="4">
Homework 1
</font>
</center>
</p>

<p>
<center>
<font size="3">
Data Science, Columbian College of Arts & Sciences, George Washington University
</font>
</center>
</p>

<p>
<center>
<font size="3">
Yuxiao Huang
</font>
</center>
</p>
</b>

# Overview

- This notebook includes homework 1 for Training Shallow Models (Part 2, Chapter 2, Section 2).
- See the accompanied slides in our [github repository](https://github.com/yuxiaohuang/teaching/tree/master/gwu/machine_learning_I/summer_2023/slides/p2_shallow_learning/p2_c2_supervised_learning/p2_c2_s2_training_shallow_models).
- Here we will work on kaggle competation [New York City Taxi Trip Duration](https://www.kaggle.com/c/nyc-taxi-trip-duration/overview).
- The goal of this homework is tweaking the pipeline (including data preprocessing, hyperparameter tuning and model selection) implemented in [/p2 c2 s2 training shallow models/case_study](https://github.com/yuxiaohuang/teaching/blob/master/gwu/machine_learning_I/summer_2023/code/p2_shallow_learning/p2_c2_supervised_learning/p2_c2_s2_training_shallow_models/case_study/case_study.ipynb) to make it work for the new kaggle competation mentioned above.
- Complete the missing parts indicated by # Implement me.
- Particularly, the code should
    - be bug-free (note that the output produced by your solution may not necessarily be the same as the provided output, due to version issues)
    - be commented
- **The baseline for this assignment is -0.332515. As per the grading rubrics (section 15 of the syllabus), if the validation score of your best model is lower than or the same as the baseline, you will only receive 80% of the full mark. See more details in the grading rubrics.**
- Submit an ipynb file named homework_1.ipynb to [blackboard](https://blackboard.gwu.edu) folder /Assignments/Homework_1/.

# Notebook Configuration

## Google drive

In [ ]:
from google.colab import drive
import sys

# Mount Google Drive
drive.mount('/content/drive')

# Get the absolute path of the current folder
abspath_curr = '/content/drive/My Drive/Colab Notebooks/teaching/gwu/machine_learning_I/homework/homework_1/'

# Get the absolute path of the shallow utilities folder
abspath_util_shallow = '/content/drive/My Drive/Colab Notebooks/teaching/gwu/machine_learning_I/code/utilities/p2_shallow_learning/'

# Get the absolute path of the shallow models folder
abspath_model_shallow = '/content/drive/My Drive/Colab Notebooks/teaching/gwu/machine_learning_I/code/models/p2_shallow_learning/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Warning

In [ ]:
import warnings

# Ignore warnings
warnings.filterwarnings('ignore')

## Matplotlib

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

# Set matplotlib sizes
plt.rc('font', size=20)
plt.rc('axes', titlesize=20)
plt.rc('axes', labelsize=20)
plt.rc('xtick', labelsize=20)
plt.rc('ytick', labelsize=20)
plt.rc('legend', fontsize=20)
plt.rc('figure', titlesize=20)

## TensorFlow

In [ ]:
# The magic below allows us to use tensorflow version 2.x
%tensorflow_version 2.x
import tensorflow as tf
from tensorflow import keras

Colab only includes TensorFlow 2.x; %tensorflow_version has no effect.


## Random seed

In [ ]:
# The random seed
random_seed = 42

# Set random seed in tensorflow
tf.random.set_seed(random_seed)

# Set random seed in numpy
import numpy as np
np.random.seed(random_seed)

# Data Preprocessing
**Note: You will need to specify the date time variables, *datetime_vars*, when handling date time variables**.

In [ ]:
# Change working directory to the absolute path of the shallow utilities folder
%cd $abspath_util_shallow

# Import the shallow utitilities
%run pmlm_utilities_shallow.ipynb

/content/drive/My Drive/Colab Notebooks/teaching/gwu/machine_learning_I/code/utilities/p2_shallow_learning


## Loading the data

In [ ]:


import pandas as pd

# Load the raw training data
df_raw_train = pd.read_csv(abspath_curr + 'data/taxi/train.csv',
                           header=0)
# Make a copy of df_raw_train
df_train = df_raw_train.copy(deep=True)

# Load the raw test data
df_raw_test = pd.read_csv(abspath_curr + 'data/taxi/test.csv',
                          header=0)
# Make a copy of df_raw_test
df_test = df_raw_test.copy(deep=True)

# Get the name of the target
target = 'trip_duration'



In [ ]:


# Print the dimension of df_train
pd.DataFrame([[df_train.shape[0], df_train.shape[1]]], columns=['# rows', '# columns'])



# rows  # columns
0  1458644         11

In [ ]:
# Print the dimension of df_test
pd.DataFrame([[df_test.shape[0], df_test.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0  625134          9

In [ ]:
# Print the first 5 rows of df_train
df_train.head()

id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude store_and_fwd_flag  trip_duration  
0         40.765602                  N            455  
1         40.731152                  N            663  
2         40.710087                  N           2124  
3         40.706718                  N            429  
4         40.782520                  N            435

In [ ]:


# Print the first 5 rows of df_test
df_test.head()



id  vendor_id      pickup_datetime  passenger_count  \
0  id3004672          1  2016-06-30 23:59:58                1   
1  id3505355          1  2016-06-30 23:59:53                1   
2  id1217141          1  2016-06-30 23:59:47                1   
3  id2150126          2  2016-06-30 23:59:41                1   
4  id1598245          1  2016-06-30 23:59:33                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.988129        40.732029         -73.990173         40.756680   
1        -73.964203        40.679993         -73.959808         40.655403   
2        -73.997437        40.737583         -73.986160         40.729523   
3        -73.956070        40.771900         -73.986427         40.730469   
4        -73.970215        40.761475         -73.961510         40.755890   

  store_and_fwd_flag  
0                  N  
1                  N  
2                  N  
3                  N  
4                  N

In [ ]:


from sklearn.model_selection import train_test_split

# Divide the training data into training (80%) and validation (20%)
df_train, df_val = train_test_split(df_train, train_size=0.8, random_state=random_seed)

# Reset the index
df_train, df_val = df_train.reset_index(drop=True), df_val.reset_index(drop=True)



In [ ]:
# Print the dimension of df_train
pd.DataFrame([[df_train.shape[0], df_train.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0  1166915         11

In [ ]:


# Print the dimension of df_val
pd.DataFrame([[df_val.shape[0], df_val.shape[1]]], columns=['# rows', '# columns'])



# rows  # columns
0  291729         11

## Handling uncommon features

In [ ]:


# Call common_var_checker
# See the implementation in pmlm_utilities.ipynb
df_common_var = common_var_checker(df_train, df_val, df_test, target)

# Print df_common_var
df_common_var



common var
0    dropoff_latitude
1   dropoff_longitude
2                  id
3     passenger_count
4     pickup_datetime
5     pickup_latitude
6    pickup_longitude
7  store_and_fwd_flag
8       trip_duration
9           vendor_id

In [ ]:
# Get the features in the training data but not in the validation or test data
uncommon_feature_train_not_val_test = np.setdiff1d(df_train.columns, df_common_var['common var'])

# Print the uncommon features
pd.DataFrame(uncommon_feature_train_not_val_test, columns=['uncommon feature'])

uncommon feature
0  dropoff_datetime

In [ ]:
# Get the features in the validation data but not in the training or test data
uncommon_feature_val_not_train_test = np.setdiff1d(df_val.columns, df_common_var['common var'])

# Print the uncommon features
pd.DataFrame(uncommon_feature_val_not_train_test, columns=['uncommon feature'])

uncommon feature
0  dropoff_datetime

In [ ]:
# Get the features in the test data but not in the training or validation data
uncommon_feature_test_not_train_val = np.setdiff1d(df_test.columns, df_common_var['common var'])

# Print the uncommon features
pd.DataFrame(uncommon_feature_test_not_train_val, columns=['uncommon feature'])

Empty DataFrame
Columns: [uncommon feature]
Index: []

## Removing uncommon features

In [ ]:


# Remove the uncommon features from the training data
df_train = df_train.drop(columns=uncommon_feature_train_not_val_test)

# Print the first 5 rows of df_train
df_train.head()



id  vendor_id      pickup_datetime  passenger_count  \
0  id0458976          2  2016-06-29 18:21:02                1   
1  id0434613          2  2016-04-25 13:03:26                1   
2  id3809234          2  2016-05-07 12:36:09                1   
3  id1203705          1  2016-05-14 18:44:17                1   
4  id1896645          2  2016-04-10 22:51:25                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.862762        40.768822         -73.891701         40.746689   
1        -73.958038        40.783237         -73.975510         40.760853   
2        -73.969460        40.785519         -73.989243         40.771748   
3        -73.981743        40.736549         -73.998352         40.726440   
4        -73.977913        40.752609         -73.975647         40.733139   

  store_and_fwd_flag  trip_duration  
0                  N           1133  
1                  N            887  
2                  N            686  
3                  N            818  
4                  N            951

In [ ]:


# Remove the uncommon features from the validation data
df_val = df_val.drop(columns=uncommon_feature_val_not_train_test)

# Print the first 5 rows of df_val
df_val.head()



id  vendor_id      pickup_datetime  passenger_count  \
0  id2793718          2  2016-06-08 07:36:19                1   
1  id3485529          2  2016-04-03 12:58:11                1   
2  id1816614          2  2016-06-05 02:49:13                5   
3  id1050851          2  2016-05-05 17:18:27                2   
4  id0140657          1  2016-05-12 17:43:38                4   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.985611        40.735943         -73.980331         40.760468   
1        -73.978394        40.764351         -73.991623         40.749859   
2        -73.989059        40.744389         -73.973381         40.748692   
3        -73.990326        40.731136         -73.991264         40.748917   
4        -73.789497        40.646675         -73.987137         40.759232   

  store_and_fwd_flag  trip_duration  
0                  N           1040  
1                  N            827  
2                  N            614  
3                  N            867  
4                  N           4967

In [ ]:


# Remove the uncommon features from the test data
df_test = df_test.drop(columns=uncommon_feature_test_not_train_val)

# Print the first 5 rows of df_test
df_test.head()



id  vendor_id      pickup_datetime  passenger_count  \
0  id3004672          1  2016-06-30 23:59:58                1   
1  id3505355          1  2016-06-30 23:59:53                1   
2  id1217141          1  2016-06-30 23:59:47                1   
3  id2150126          2  2016-06-30 23:59:41                1   
4  id1598245          1  2016-06-30 23:59:33                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.988129        40.732029         -73.990173         40.756680   
1        -73.964203        40.679993         -73.959808         40.655403   
2        -73.997437        40.737583         -73.986160         40.729523   
3        -73.956070        40.771900         -73.986427         40.730469   
4        -73.970215        40.761475         -73.961510         40.755890   

  store_and_fwd_flag  
0                  N  
1                  N  
2                  N  
3                  N  
4                  N

## Handling identifiers

In [ ]:


# Combine df_train, df_val and df_test
df = pd.concat([df_train, df_val, df_test], sort=False)



In [ ]:


# Call id_checker on df
# See the implementation in pmlm_utilities.ipynb
df_id = id_checker(df)

# Print the first 5 rows of df_id
df_id.head()



id
0  id0458976
1  id0434613
2  id3809234
3  id1203705
4  id1896645

## Removing identifiers

In [ ]:


import numpy as np

# Remove identifiers from df_train
df_train.drop(columns=np.intersect1d(df_id.columns, df_train.columns), inplace=True)

# Remove identifiers from df_val
df_val.drop(columns=np.intersect1d(df_id.columns, df_val.columns), inplace=True)

# Remove identifiers from df_test
df_test.drop(columns=np.intersect1d(df_id.columns, df_test.columns), inplace=True)



In [ ]:
# Print the first 5 rows of df_train
df_train.head()

vendor_id      pickup_datetime  passenger_count  pickup_longitude  \
0          2  2016-06-29 18:21:02                1        -73.862762   
1          2  2016-04-25 13:03:26                1        -73.958038   
2          2  2016-05-07 12:36:09                1        -73.969460   
3          1  2016-05-14 18:44:17                1        -73.981743   
4          2  2016-04-10 22:51:25                1        -73.977913   

   pickup_latitude  dropoff_longitude  dropoff_latitude store_and_fwd_flag  \
0        40.768822         -73.891701         40.746689                  N   
1        40.783237         -73.975510         40.760853                  N   
2        40.785519         -73.989243         40.771748                  N   
3        40.736549         -73.998352         40.726440                  N   
4        40.752609         -73.975647         40.733139                  N   

   trip_duration  
0           1133  
1            887  
2            686  
3            818  
4            951

In [ ]:
# Print the first 5 rows of df_val
df_val.head()

vendor_id      pickup_datetime  passenger_count  pickup_longitude  \
0          2  2016-06-08 07:36:19                1        -73.985611   
1          2  2016-04-03 12:58:11                1        -73.978394   
2          2  2016-06-05 02:49:13                5        -73.989059   
3          2  2016-05-05 17:18:27                2        -73.990326   
4          1  2016-05-12 17:43:38                4        -73.789497   

   pickup_latitude  dropoff_longitude  dropoff_latitude store_and_fwd_flag  \
0        40.735943         -73.980331         40.760468                  N   
1        40.764351         -73.991623         40.749859                  N   
2        40.744389         -73.973381         40.748692                  N   
3        40.731136         -73.991264         40.748917                  N   
4        40.646675         -73.987137         40.759232                  N   

   trip_duration  
0           1040  
1            827  
2            614  
3            867  
4           4967

In [ ]:


# Print the first 5 rows of df_test
df_test.head()



vendor_id      pickup_datetime  passenger_count  pickup_longitude  \
0          1  2016-06-30 23:59:58                1        -73.988129   
1          1  2016-06-30 23:59:53                1        -73.964203   
2          1  2016-06-30 23:59:47                1        -73.997437   
3          2  2016-06-30 23:59:41                1        -73.956070   
4          1  2016-06-30 23:59:33                1        -73.970215   

   pickup_latitude  dropoff_longitude  dropoff_latitude store_and_fwd_flag  
0        40.732029         -73.990173         40.756680                  N  
1        40.679993         -73.959808         40.655403                  N  
2        40.737583         -73.986160         40.729523                  N  
3        40.771900         -73.986427         40.730469                  N  
4        40.761475         -73.961510         40.755890                  N

## Handling date/time variables

In [ ]:

# Get the date time variables
datetime_vars = ['pickup_datetime']



In [ ]:


# Call datetime_transformer on df_train
# See the implementation in pmlm_utilities.ipynb
df_train = datetime_transformer(df_train, datetime_vars)

# Print the first 5 rows of df_train
df_train.head()



vendor_id  passenger_count  pickup_longitude  pickup_latitude  \
0          2                1        -73.862762        40.768822   
1          2                1        -73.958038        40.783237   
2          2                1        -73.969460        40.785519   
3          1                1        -73.981743        40.736549   
4          2                1        -73.977913        40.752609   

   dropoff_longitude  dropoff_latitude store_and_fwd_flag  trip_duration  \
0         -73.891701         40.746689                  N           1133   
1         -73.975510         40.760853                  N            887   
2         -73.989243         40.771748                  N            686   
3         -73.998352         40.726440                  N            818   
4         -73.975647         40.733139                  N            951   

   pickup_datetime_year  pickup_datetime_month  pickup_datetime_day  \
0                  2016                      6                   29   
1                  2016                      4                   25   
2                  2016                      5                    7   
3                  2016                      5                   14   
4                  2016                      4                   10   

   pickup_datetime_hour  pickup_datetime_minute  pickup_datetime_second  
0                    18                      21                       2  
1                    13                       3                      26  
2                    12                      36                       9  
3                    18                      44                      17  
4                    22                      51                      25

In [ ]:


# Call datetime_transformer on df_val
# See the implementation in pmlm_utilities.ipynb
df_val = datetime_transformer(df_val, datetime_vars)

# Print the first 5 rows of df_val
df_val.head()



vendor_id  passenger_count  pickup_longitude  pickup_latitude  \
0          2                1        -73.985611        40.735943   
1          2                1        -73.978394        40.764351   
2          2                5        -73.989059        40.744389   
3          2                2        -73.990326        40.731136   
4          1                4        -73.789497        40.646675   

   dropoff_longitude  dropoff_latitude store_and_fwd_flag  trip_duration  \
0         -73.980331         40.760468                  N           1040   
1         -73.991623         40.749859                  N            827   
2         -73.973381         40.748692                  N            614   
3         -73.991264         40.748917                  N            867   
4         -73.987137         40.759232                  N           4967   

   pickup_datetime_year  pickup_datetime_month  pickup_datetime_day  \
0                  2016                      6                    8   
1                  2016                      4                    3   
2                  2016                      6                    5   
3                  2016                      5                    5   
4                  2016                      5                   12   

   pickup_datetime_hour  pickup_datetime_minute  pickup_datetime_second  
0                     7                      36                      19  
1                    12                      58                      11  
2                     2                      49                      13  
3                    17                      18                      27  
4                    17                      43                      38

In [ ]:
# Call datetime_transformer on df_test
# See the implementation in pmlm_utilities.ipynb
df_test = datetime_transformer(df_test, datetime_vars)

# Print the first 5 rows of df_test
df_test.head()

vendor_id  passenger_count  pickup_longitude  pickup_latitude  \
0          1                1        -73.988129        40.732029   
1          1                1        -73.964203        40.679993   
2          1                1        -73.997437        40.737583   
3          2                1        -73.956070        40.771900   
4          1                1        -73.970215        40.761475   

   dropoff_longitude  dropoff_latitude store_and_fwd_flag  \
0         -73.990173         40.756680                  N   
1         -73.959808         40.655403                  N   
2         -73.986160         40.729523                  N   
3         -73.986427         40.730469                  N   
4         -73.961510         40.755890                  N   

   pickup_datetime_year  pickup_datetime_month  pickup_datetime_day  \
0                  2016                      6                   30   
1                  2016                      6                   30   
2                  2016                      6                   30   
3                  2016                      6                   30   
4                  2016                      6                   30   

   pickup_datetime_hour  pickup_datetime_minute  pickup_datetime_second  
0                    23                      59                      58  
1                    23                      59                      53  
2                    23                      59                      47  
3                    23                      59                      41  
4                    23                      59                      33

## Handling missing data

In [ ]:


# Combine df_train, df_val and df_test
df = pd.concat([df_train, df_val, df_test], sort=False)



In [ ]:
# Call nan_checker on df
# See the implementation in pmlm_utilities.ipynb
df_nan = nan_checker(df)

# Print df_nan
df_nan


var  proportion    dtype
0  trip_duration         0.3  float64

In [ ]:
# Print the unique data type of variables with NaN
pd.DataFrame(df_nan['dtype'].unique(), columns=['dtype'])

dtype
0  float64

In [ ]:


# Get the variables with missing values, their proportion of missing values and data type
df_miss = df_nan[df_nan['dtype'] == 'float64'].reset_index(drop=True)

# Print df_miss
df_miss



var  proportion    dtype
0  trip_duration         0.3  float64

### Split train, val, and test data

In [ ]:
# Separating the training data
df_train = df.iloc[:df_train.shape[0], :]

# Separating the validation data
df_val = df.iloc[df_train.shape[0]:df_train.shape[0] + df_val.shape[0], :]

# Separating the test data
df_test = df.iloc[df_train.shape[0] + df_val.shape[0]:, :]

In [ ]:


# Print the dimension of df_train
pd.DataFrame([[df_train.shape[0], df_train.shape[1]]], columns=['# rows', '# columns'])



# rows  # columns
0  1166915         14

In [ ]:


# Print the dimension of df_val
pd.DataFrame([[df_val.shape[0], df_val.shape[1]]], columns=['# rows', '# columns'])



# rows  # columns
0  291729         14

In [ ]:


# Print the dimension of df_test
pd.DataFrame([[df_test.shape[0], df_test.shape[1]]], columns=['# rows', '# columns'])



# rows  # columns
0  625134         14

### Imputing missing values

In [ ]:
from sklearn.impute import SimpleImputer

# If there are missing values
if len(df_miss['var']) > 0:
    # The SimpleImputer
    si = SimpleImputer(missing_values=np.nan, strategy='mean')

    # Impute the variables with missing values in df_train, df_val and df_test
    df_train[df_miss['var']] = si.fit_transform(df_train[df_miss['var']])
    df_val[df_miss['var']] = si.transform(df_val[df_miss['var']])
    df_test[df_miss['var']] = si.transform(df_test[df_miss['var']])

### Coding the values

In [ ]:
# Combine df_train, df_val and df_test
df = pd.concat([df_train, df_val, df_test], sort=False)

# Print the unique data type of variables in df
pd.DataFrame(df.dtypes.unique(), columns=['dtype'])

dtype
0    int64
1  float64
2   object

### Identify categoricals

In [ ]:
# Call cat_var_checker on df
# See the implementation in pmlm_utilities.ipynb
df_cat = cat_var_checker(df)

# Print the dataframe
df_cat

var  nunique
0  store_and_fwd_flag        2

### One hot encode categoricals

In [ ]:
# One-hot-encode the categorical features in the combined data
df = pd.get_dummies(df, columns=np.setdiff1d(df_cat['var'], [target]))

# Print the first 5 rows of df
df.head()

vendor_id  passenger_count  pickup_longitude  pickup_latitude  \
0          2                1        -73.862762        40.768822   
1          2                1        -73.958038        40.783237   
2          2                1        -73.969460        40.785519   
3          1                1        -73.981743        40.736549   
4          2                1        -73.977913        40.752609   

   dropoff_longitude  dropoff_latitude  trip_duration  pickup_datetime_year  \
0         -73.891701         40.746689         1133.0                  2016   
1         -73.975510         40.760853          887.0                  2016   
2         -73.989243         40.771748          686.0                  2016   
3         -73.998352         40.726440          818.0                  2016   
4         -73.975647         40.733139          951.0                  2016   

   pickup_datetime_month  pickup_datetime_day  pickup_datetime_hour  \
0                      6                   29                    18   
1                      4                   25                    13   
2                      5                    7                    12   
3                      5                   14                    18   
4                      4                   10                    22   

   pickup_datetime_minute  pickup_datetime_second  store_and_fwd_flag_N  \
0                      21                       2                     1   
1                       3                      26                     1   
2                      36                       9                     1   
3                      44                      17                     1   
4                      51                      25                     1   

   store_and_fwd_flag_Y  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

### Split train, val, and test again

In [ ]:
# Separating the training data
df_train = df.iloc[:df_train.shape[0], :]

# Separating the validation data
df_val = df.iloc[df_train.shape[0]:df_train.shape[0] + df_val.shape[0], :]

# Separating the test data
df_test = df.iloc[df_train.shape[0] + df_val.shape[0]:, :]

In [ ]:
# Print the dimension of df_train
pd.DataFrame([[df_train.shape[0], df_train.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0  1166915         15

In [ ]:
# Print the dimension of df_val
pd.DataFrame([[df_val.shape[0], df_val.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0  291729         15

In [ ]:
# Print the dimension of df_test
pd.DataFrame([[df_test.shape[0], df_test.shape[1]]], columns=['# rows', '# columns'])

# rows  # columns
0  625134         15

## Split features and target

In [ ]:
# Get the feature matrix
X_train = df_train[np.setdiff1d(df_train.columns, [target])].values
X_val = df_val[np.setdiff1d(df_val.columns, [target])].values
X_test = df_test[np.setdiff1d(df_test.columns, [target])].values

# Get the target vector
y_train = df_train[target].values
y_val = df_val[target].values
y_test = df_test[target].values

# Scale data
## Standardize

In [ ]:


from sklearn.preprocessing import StandardScaler

# The StandardScaler
ss = StandardScaler()



### Features

In [ ]:
# Standardize the training data
X_train = ss.fit_transform(X_train)

# Standardize the validation data
X_val = ss.transform(X_val)

# Standardize the test data
X_test = ss.transform(X_test)

### Target

In [ ]:
# Standardize the training data
y_train = ss.fit_transform(y_train.reshape(-1, 1)).reshape(-1)

# Standardize the validation data
y_val = ss.transform(y_val.reshape(-1, 1)).reshape(-1)

# Standardize the test data
y_test = ss.transform(y_test.reshape(-1, 1)).reshape(-1)

# Hyperparameter Tuning

## Create models dictionary

In [ ]:


from sklearn.linear_model import SGDRegressor

models = {'sgd': SGDRegressor(random_state=random_seed)}



## Create pipelines dictionary

In [ ]:
from sklearn.pipeline import Pipeline

pipes = {}

for acronym, model in models.items():
    pipes[acronym] = Pipeline([('model', model)])

## Obtain predefined split crossvalidator

In [ ]:
# Get the:
# feature matrix and target velctor in the combined training and validation data
# target vector in the combined training and validation data
# PredefinedSplit
# See the implementation in pmlm_utilities.ipynb
X_train_val, y_train_val, ps = get_train_val_ps(X_train, y_train, X_val, y_val)

## GridsearchCV

### Create dictionary of parameter grids

In [ ]:
param_grids = {}

### Parameter grad for SGDregressor

In [ ]:
# The parameter grid of eta
eta_grid = [0.001, 0.01]

# The parameter grid of alpha
alpha_grid = [0.01, 0.1]

# Update param_grids
param_grids['sgd'] = [{'model__eta0': eta_grid,
                       'model__alpha': alpha_grid}]

### Create directory for CV results

In [ ]:
# Make directory
directory = os.path.dirname(abspath_curr + 'result/cv_results/GridSearchCV/')
if not os.path.exists(directory):
    os.makedirs(directory)

### Tune hyperparameters

In [ ]:
from sklearn.model_selection import GridSearchCV

# The list of [best_score_, best_params_, best_estimator_] obtained by GridSearchCV
best_score_params_estimator_gs = []

# For each model
for acronym in pipes.keys():
    # GridSearchCV
    gs = GridSearchCV(estimator=pipes[acronym],
                      param_grid=param_grids[acronym],
                      scoring='neg_mean_squared_error',
                      n_jobs=2,
                      cv=ps,
                      return_train_score=True)

    # Fit the pipeline
    gs = gs.fit(X_train_val, y_train_val)

    # Update best_score_params_estimator_gs
    best_score_params_estimator_gs.append([gs.best_score_, gs.best_params_, gs.best_estimator_])

    # Sort cv_results in ascending order of 'rank_test_score' and 'std_test_score'
    cv_results = pd.DataFrame.from_dict(gs.cv_results_).sort_values(by=['rank_test_score', 'std_test_score'])

    # Get the important columns in cv_results
    important_columns = ['rank_test_score',
                         'mean_test_score',
                         'std_test_score',
                         'mean_train_score',
                         'std_train_score',
                         'mean_fit_time',
                         'std_fit_time',
                         'mean_score_time',
                         'std_score_time']

    # Move the important columns ahead
    cv_results = cv_results[important_columns + sorted(list(set(cv_results.columns) - set(important_columns)))]

    # Write cv_results file
    cv_results.to_csv(path_or_buf=abspath_curr + 'result/cv_results/GridSearchCV/' + acronym + '.csv', index=False)

# Sort best_score_params_estimator_gs in descending order of the best_score_
best_score_params_estimator_gs = sorted(best_score_params_estimator_gs, key=lambda x : x[0], reverse=True)

# Print best_score_params_estimator_gs
pd.DataFrame(best_score_params_estimator_gs, columns=['best_score', 'best_param', 'best_estimator'])

best_score                                  best_param  \
0   -0.332515  {'model__alpha': 0.1, 'model__eta0': 0.01}   

                               best_estimator  
0  (SGDRegressor(alpha=0.1, random_state=42))

## RandomizedSearchCV
### Create dictionary of parameter distributions

In [ ]:
param_dists = {}

In [ ]:
from scipy.stats import uniform

# The distribution for eta: a uniform distribution over [loc, loc + scale]
eta_dist = uniform(loc=0.001, scale=0.019)

# The distribution for alpha: a uniform distribution over [loc, loc + scale]
alpha_dist = uniform(loc=0.01, scale=0.09)

# Update param_dists
param_dists['sgd'] = [{'model__eta0': eta_dist,
                       'model__alpha': alpha_dist}]

In [ ]:
# The distribution for eta: a uniform distribution over [loc, loc + scale]
eta_dist = uniform(loc=0.001, scale=0.019)

# The distribution for alpha: a uniform distribution over [loc, loc + scale]
alpha_dist = uniform(loc=0.01, scale=0.09)

# Update param_dists
param_dists['lr_mbgd'] = [{'model__eta': eta_dist,
                           'model__alpha': alpha_dist}]

### Create directory for CV results

In [ ]:
# Make directory
directory = os.path.dirname(abspath_curr + 'result/cv_results/RandomizedSearchCV/')
if not os.path.exists(directory):
    os.makedirs(directory)

### Tune hyperparameters

In [ ]:


from sklearn.model_selection import RandomizedSearchCV

# The list of [best_score_, best_params_, best_estimator_] obtained by RandomizedSearchCV
best_score_params_estimator_rs = []

for acronym in pipes.keys():
    # RandomizedSearchCV
    rs = RandomizedSearchCV(estimator=pipes[acronym],
                            param_distributions=param_dists[acronym],
                            n_iter=4,
                            scoring='neg_mean_squared_error',
                            n_jobs=2,
                            cv=ps,
                            random_state=random_seed,
                            return_train_score=True)

    # Fit the pipeline
    rs = rs.fit(X_train_val, y_train_val)

    # Update best_score_param_estimators
    best_score_params_estimator_rs.append([rs.best_score_, rs.best_params_, rs.best_estimator_])

    # Sort cv_results in ascending order of 'rank_test_score' and 'std_test_score'
    cv_results = pd.DataFrame.from_dict(rs.cv_results_).sort_values(by=['rank_test_score', 'std_test_score'])

    # Get the important columns in cv_results
    important_columns = ['rank_test_score',
                         'mean_test_score',
                         'std_test_score',
                         'mean_train_score',
                         'std_train_score',
                         'mean_fit_time',
                         'std_fit_time',
                         'mean_score_time',
                         'std_score_time']

    # Move the important columns ahead
    cv_results = cv_results[important_columns + sorted(list(set(cv_results.columns) - set(important_columns)))]

    # Write cv_results file
    cv_results.to_csv(path_or_buf=abspath_curr + 'result/cv_results/RandomizedSearchCV/' + acronym + '.csv', index=False)

# Sort best_score_params_estimator_rs in descending order of the best_score_
best_score_params_estimator_rs = sorted(best_score_params_estimator_rs, key=lambda x : x[0], reverse=True)

# Print best_score_params_estimator_rs
pd.DataFrame(best_score_params_estimator_rs, columns=['best_score', 'best_param', 'best_estimator'])



best_score                                         best_param  \
0   -0.332487  {'model__alpha': 0.07587945476302645, 'model__...   

                                      best_estimator  
0  (SGDRegressor(alpha=0.07587945476302645, eta0=...

# Model Selection

After hitting the provided baseline score with gridsearchcv (-0.332515), we then beat it with the best model from randomizedsearchcv (-0.332487). We will therefore select that model.

In [ ]:
# Get the best_score, best_params and best_estimator obtained by RandomizedSearchCV
best_score_rs, best_params_rs, best_estimator_rs = best_score_params_estimator_rs[0]


# Generating the Submission File
Use the best model selected earlier to generate the submission file for this kaggle competition.

## Creating the directory for the submission file

In [ ]:
# Make directory
directory = os.path.dirname(abspath_curr + 'result/submission/')
if not os.path.exists(directory):
    os.makedirs(directory)

## Generating the submission file

In [ ]:
# Get the prediction on the testing data using best_model
y_test_pred = best_estimator_rs.predict(X_test)

# Get the dataframe of y_test_pred, which has the same shape as df_train
df_y_test_pred = pd.DataFrame(np.tile(y_test_pred.reshape(-1, 1), df_train.shape[1]),
                              columns=df_train.columns)

# Transform df_y_test_pred back to the original scale
df_y_test_pred = pd.DataFrame(ss.inverse_transform(df_y_test_pred),
                              columns=df_train.columns)

# Get the submission dataframe
df_submit = pd.DataFrame(np.hstack((df_raw_test[['id']], df_y_test_pred[[target]])),
                         columns=['id', target])

# Generate the submission file
df_submit.to_csv(abspath_curr + 'result/submission/submission.csv', index=False)